### PR TITLE
Feature/36409 analyse files for anomaly report

### PR DESCRIPTION
**Pull request by jon-shipley**
Refactor anomaly report to be run from file

First two detections

### CODE REVIEW

File: functions-app/package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "license": "GPL-3.0",
   "scripts": {
     "start": "func host start --port 7072",
-    "test": "yarn lint && yarn run nyc -r html jasmine **/*.spec.js */**/*.spec.js",
+    "test": "yarn lint && yarn run nyc -r html jasmine **/*.spec.js */**/*.spec.js */**/**/*.spec.js",
```
Comment (GuyHarwood) [inline functions-app/package.json]: would `**/*.spec.js` not cover all your subdirectories in this particular case or am i overlooking something subtle?

File: functions-app/package.json [hunk 2]
```diff
@@ -25,6 +25,7 @@
     "mssql": "^5.1.0",
     "object-hash": "^1.3.1",
     "ramda": "^0.26.1",
+    "ramda-adjunct": "^2.20.0",
```
Comment (GeorgeChatzigiannis) [inline functions-app/package.json]: do we really need this package? methods used can be replaced with vanilla ones?

File: functions-app/report-psychometrician/service/detections/get-check-started-date.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+const R = require('ramda')
+
+function getCheckStartedDateFromAuditLog (payload) {
+  if (typeof payload !== 'object') return undefined
+  const audits = R.prop('audit', payload)
+  if (!audits) return undefined
+  const checkStartedEvent = R.find(R.propEq('type', 'CheckStarted'), audits)
+  if (!checkStartedEvent) return undefined
+  const ts = R.prop('clientTimestamp', checkStartedEvent)
+  if (!ts) return undefined
```
Comment (GeorgeChatzigiannis) [inline functions-app/report-psychometrician/service/detections/get-check-started-date.js]: is specifically undefined required here as a returned value?

File: functions-app/report-psychometrician/service/detections/remove-duplicate-anomaly-reports.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * For example, perhaps because multiple inputs were recorded before the question was shown
+ * @usage removeDuplicates(anomalyReports)
+ */
+const removeDuplicates = R.reduce(
+  (accumulator, newElement) => {
+    const props = ['Question number', 'Message']
+    const found = R.filter(e => R.equals(
+      R.pick(props, e),
+      R.pick(props, newElement)),
+    accumulator)
```
Comment (GeorgeChatzigiannis) [inline functions-app/report-psychometrician/service/detections/remove-duplicate-anomaly-reports.js]: 🏅

File: functions-app/package.json (hunk 1 above)
Comment (jon-shipley) [reply functions-app/package.json]: It was needed as due to the extra level of sub-directory nesting with the 'detections' directory.  So, apparently not.

Comment (GuyHarwood) [reply functions-app/package.json]: that's odd.  i was under the impression that the first one `**/*.js` covers all recursively

File: functions-app/package.json (hunk 2 above)
Comment (jon-shipley) [reply functions-app/package.json]: I suspect the answer, at the moment, is that we could leave it behind. And yet, it offers such a wide range of useful ancillary functions to Ramda that it seems logical to include the library, especially as we get such great use from ramda for the processing piece.

We have many more detections to add, these will most likely make use of ramda so I want to include at this time.  Please note that ramda-adjunct does not have any dependencies other than ramda.

File: functions-app/report-psychometrician/service/detections/get-check-started-date.js (hunk 3 above)
Comment (jon-shipley) [reply functions-app/report-psychometrician/service/detections/get-check-started-date.js]: What would you suggest?

File: functions-app/package.json (hunk 2 above)
Comment (GeorgeChatzigiannis) [reply functions-app/package.json]: Ok then it does make sense to keep it.

File: functions-app/report-psychometrician/service/detections/get-check-started-date.js (hunk 3 above)
Comment (GeorgeChatzigiannis) [reply functions-app/report-psychometrician/service/detections/get-check-started-date.js]: Wouldn't plain return be sufficient?I'd expect the caller to handle it properly.

Comment (jon-shipley) [reply functions-app/report-psychometrician/service/detections/get-check-started-date.js]: Adjusted as you suggest :)

**Pull request merged**